src/Smearing.jl

```diff
@@ -93,47 +96,50 @@ function occupation_divided_difference(S::FermiDirac, x, y, εF, temp)
 end
 
 struct Gaussian <: SmearingFunction end
-occupation(S::Gaussian, x) = (1 - erf(x)) / 2
-entropy(S::Gaussian, x) = 1 / (2sqrt(typeof(x)(pi))) * exp(-x^2)
+occupation(S::Gaussian, x) = erfc(x) / 2
+entropy(S::Gaussian, x::T) where T = 1 / (2 * sqrt(T(pi))) * exp(-x^2)
 
-H1(x) = 2x
-H2(x) = 4x^2 - 2
-H3(x) = 8x^3 - 12x
-H4(x) = 16x^4 - 48x^2 + 12
-A(n, T=Float64) = (-1)^n / (factorial(n) * 4^n * sqrt(T(π)))
+struct MarzariVanderbilt <: SmearingFunction end
+occupation(S::MarzariVanderbilt, x::T) where T = erf(-x - 1/sqrt(T(2))) / 2 + 1/sqrt(2*T(π)) * exp(-(-x - 1/sqrt(T(2)))^2) + 1/T(2)
+entropy(S::MarzariVanderbilt, x::T) where T = -1/sqrt(2*T(π)) * (-x - 1/sqrt(T(2))) * exp(-(-x - 1/sqrt(T(2)))^2)
 
-struct MethfesselPaxton1 <: SmearingFunction end
-function occupation(S::MethfesselPaxton1, x)
-    x == Inf && return zero(x)
-    x == -Inf && return one(x)
-    occupation(Gaussian(), x) + A(1, typeof(x))*H1(x)*exp(-x^2)
+A(T, n) = (-1)^n / (factorial(n) * 4^n * sqrt(T(π)))
+"""
+Standard Hermite function using physicist's convention.
+"""
+function H(x, n)
+    if n == 0
+        return one(x)
+    elseif n == 1
+        return 2 * x
+    else
+        return 2 * x * H(x, n-1) - 2 * (n-1) * H(x, n-2)
+    end
 end
-entropy(S::MethfesselPaxton1, x) = 1/2 * A(1, typeof(x)) * H2(x) * exp(-x^2)
 
-struct MethfesselPaxton2 <: SmearingFunction end
-function occupation(S::MethfesselPaxton2, x)
+struct MethfesselPaxton <: SmearingFunction
+    order::Int
+end
+function occupation(S::MethfesselPaxton, x::T) where T
     x == Inf && return zero(x)
     x == -Inf && return one(x)
-    occupation(MethfesselPaxton1(), x) + A(2, typeof(x))*H3(x)*exp(-x^2)
+    S₀ = erfc(x) / 2
+    ΣSₙ = zero(x)
+    for i in 1:S.order
+        ΣSₙ += A(T, i) * H(x, 2i - 1)
+    end
+    return S₀ + ΣSₙ * exp(-x^2)
 end
-entropy(S::MethfesselPaxton2, x) = 1/2 * A(2, typeof(x)) * H4(x) * exp(-x^2)
-
-function MethfesselPaxton(order::Integer)
-    if order == 0
-        Gaussian()
-    elseif order == 1
-        MethfesselPaxton1()
-    elseif order == 2
-        MethfesselPaxton2()
-    else
-        error("Not implemented")
+function entropy(S::MethfesselPaxton, x::T) where T
+    exp_term = exp(-x^2)
+    S₀ = A(T, 0) / 2 * exp_term
```

Comment:
just write it out explicitly. Hopefully the compiler is suffiicently smart (or will become sufficiently smart) to recognize `exp` is pure, but even if not this is not performance critical